js/ui/workspaceMonitor.js

```diff
@@ -10,276 +10,142 @@ const WorkspaceMonitor = new Lang.Class({
             this._updateOverview();
         }
     },
 
-    _untrackWorkspace: function() {
-        if (this._windowAddedId > 0) {
-            this._activeWorkspace.disconnect(this._windowAddedId);
-            this._windowAddedId = 0;
-        }
+    _minimizeWindow: function(shellwm, actor) {
+        let app = this._windowTracker.get_window_app(actor.meta_window);
 
-        if (this._windowRemovedId > 0) {
-            this._activeWorkspace.disconnect(this._windowRemovedId);
-            this._windowRemovedId = 0;
+        if (app in this._trackedApps) {
+            this._setAppVisible(app, this._appHasVisibleWindows(app));
+            if (this._visibleApps.size == 0) {
+                Main.layoutManager.prepareForOverview();
+            }
         }
-
-        this._activeWorkspace = null;
-
-        this._visibleWindows = 0;
-
-        this._minimizedWindows = new Hash.Map();
-        this._knownWindows = new Hash.Map();
     },
 
-    _workspaceSwitched: function(from, to, direction) {
-        this._untrackWorkspace();
-        this._trackWorkspace(this._metaScreen.get_active_workspace(), true);
-    },
+    _minimizeWindowCompleted: function(shellwm, actor) {
+        let app = this._windowTracker.get_window_app(actor.meta_window);
 
-    // The sequence of events is
-    // Window is mapped but not yet in system: _windowAdded
-    // Window is newly created: _windowAdded -> _mapWindow
-    // Window is minimized: _minimizeWindow
-    // Window is unminimized: _mapWindow
-    // Window is closed: _destroyWindow -> _windowRemoved
-    // Minimized window is closed: _windowRemoved
-    //
-    // This allows us to separate the work
-    // _windowAdded: Handles tracking of known windows
-    //               checks if a window is interesting and adds it to our list
-    //               of known windows. Also checks if it is minimized and if so
-    //               adds it to the list of minimized windows
-    // _windowRemoved: Is the opposite of _windowAdded
-    //
-    // _mapWindow: Handles tracking of visible windows
-    // _minimizeWindow:
-    // _destroyWindow: These are the opposite of _mapWindow for the 2 different
-    //                 ways a window can be taken off the screen: minimizing and
-    //                 closing
-    _realWindowAdded: function(metaWindow) {
-        // We special-case eos-speedwagon here, as we don't want to go
-        // back to the overview if the only window that was dismissed was the
-        // splash screen.
-        // See also shell_app_is_interesting_window() in src/shell-app.c, which
-        // was added for the same purpose.
-        // This code eventually should be changed to track applications instead
-        // of windows.
-        if (!Shell.WindowTracker.is_window_interesting(metaWindow) ||
-            metaWindow.get_role() == 'eos-speedwagon') {
-            return false;
+        if (app in this._trackedApps) {
```

Comment:
Not entirely sure if this is needed here, since the same block in `_minimizeWindow` should already have run for the same window.